yourbench/utils/inference_engine.py

```diff
@@ -1,122 +1,301 @@
+        return super().set_cache(new_key, value, **kwargs)
+
+    def _get_event_loop_aware_cache_key(self, key):
+        loop = asyncio.get_running_loop()
+        if not hasattr(loop, "_custom_identifier"):
+            loop._custom_identifier = uuid.uuid4()
+        loop_id = loop._custom_identifier
+        return f"{key}-{loop_id}"
+
+
+litellm.in_memory_llm_clients_cache = EventLoopAwareCache()
+
+
+@dataclass
+class Model:
+    model_name: str
+    request_style: str = "openai"
+    base_url: str = None
+    api_key: str = None
+    max_concurrent_requests: int = 8
+    inference_backend: str = "litellm"
+    provider: str = None
+    api_version: str = None
+
+
+@dataclass
+class InferenceCall:
+    messages: List[Dict[str, str]]
+    temperature: Optional[float] = None
+    tags: List[str] = field(default_factory=lambda: ["dev"])
+    max_retries: int = 8
+    seed: Optional[int] = None
+
+
+@dataclass
+class InferenceJob:
+    inference_calls: List[InferenceCall]
+
+
+@asynccontextmanager
+async def _timeout_context(timeout_seconds: int):
+    """
+    Async context manager to raise a TimeoutError if code inside takes
+    longer than 'timeout_seconds'.
+    """
+    try:
+        async with aio_timeout(timeout_seconds):
+            yield
+    except asyncio.TimeoutError:
+        logger.error("Operation timed out after {} seconds", timeout_seconds)
+        raise
+
+
+async def _get_response(model: Model, inference_call: InferenceCall) -> str:
+    """
+    Send one inference call to the model endpoint within a global timeout context.
+    Logs start/end times for better concurrency tracing.
+    """
+    start_time = time.time()
+    logger.debug("START _get_response: model='{}'  (timestamp={:.4f})", model.model_name, start_time)
+
+    if model.inference_backend == "litellm":
+        logger.debug("Litellm inference call: {}".format(inference_call.messages))
+        response = await litellm.acompletion(
+            model=f"{model.request_style}/{model.model_name}",
+            base_url=model.base_url,
+            api_key=model.api_key,
+            api_version=str(model.api_version),
+            messages=inference_call.messages,
+            temperature=inference_call.temperature,
+            metadata={"tags": inference_call.tags},
+            timeout=GLOBAL_TIMEOUT,
+        )
+    elif model.inference_backend == "hf_hub":
+        logger.debug("HF Hub inference call: {}".format(inference_call.messages))
+        # make the client first
+        # TODO: support langfuse logging with hf_hub
+        client = AsyncInferenceClient(
+            model=model.model_name,
+            token=model.api_key,
+            provider=model.provider,
+            base_url=model.base_url,
+            timeout=GLOBAL_TIMEOUT,
+        )
+        # get the response
+        response = await client.chat.completions.create(
+            messages=inference_call.messages,
+            temperature=inference_call.temperature,
+        )
```

Comment:
Need to catch if the user used another not supported backend

Reply:
if/elif/else raise or warn, depending on wyt

Reply:
threw an error - i don't plan on supporting any other backends as litellm already takes care of this